docs/design-documents/20221028-dead-letter-queue.md

```diff
@@ -0,0 +1,144 @@
+# Dead letter queue
+
+## Introduction
+
+In Conduit the concept of a Dead letter queue (DLQ) describes the place where
+messages are sent if they are negatively acknowledged (nacked). This can happen
+for a number of reasons:
+
+- A node in the pipeline experiences an error while processing a message (e.g.
+  processor).
+- The destination connector fails to process a message and sends back a nack.
+- The pipeline is stopped forcefully causing in-flight messages to be nacked.
+
+### Goals
+
+The purpose of a DLQ is to make a pipeline resilient to errors that might
+happen while processing messages.
+
+The design for the DLQ should support the following:
+
+- Allow user to reroute nacked messages to any Conduit connector (builtin or
+  standalone).
+- A message sent to the DLQ should reflect the contents it had just before the
+  error happened. Example: if a message went through 2 processors and failed
+  while being processed by a 3rd processor, the DLQ message should reflect any
+  transforms done by the first 2 processors.
+- Ensure message order is preserved when messages are sent to the DLQ.
+  In other words - the order in which messages were produced by the source
+  connector should be preserved, even if messages are nacked in a different
```

Comment:
how do we ensure this in the implementations below?

Reply:
This is actually already ensured in the source acker node using the [semaphore](https://github.com/ConduitIO/conduit/blob/b6325584d51ec64bc9086faddcdaa2788a5dfa8f/pkg/pipeline/stream/source_acker.go#L35-L37).